include/RAJA/pattern/forall.hpp

```diff
@@ -589,6 +589,136 @@ RAJA_INLINE camp::resources::EventProxy<Res> CallForallIcount::operator()(T cons
+  invoke_forall(RAJA::resources::Resource r, SEGMENT const &seg, BODY const &body)
+  {
+#if defined(RAJA_CUDA_ACTIVE)
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Cuda>(), seg, body);
+#endif
+
+#if defined(RAJA_HIP_ACTIVE)
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Hip>(), seg, body);
+#endif
+
+    return resources::EventProxy<resources::Resource>(r);
+  }
+#endif
+};
+
+template<typename t_pol, typename SEGMENT, typename BODY>
+struct IsHostResource<RAJA::resources::Host, t_pol, SEGMENT, BODY> {
+static const bool value = true;
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, SEGMENT const &seg, BODY const &body)
+  {
+    RAJA::forall<t_pol>(r.get<RAJA::resources::Host>(), seg, body);
+    return resources::EventProxy<resources::Resource>(r);
+  }
+};
+
+template<camp::idx_t IDX, typename POLICY_LIST>
+struct dynamic_helper
+{
+  template<typename SEGMENT, typename BODY>
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, const int pol, SEGMENT const &seg, BODY const &body)
+  {
+
+    using t_pol = typename camp::at<POLICY_LIST,camp::num<IDX>>::type;
+
+    if(IDX==pol){
+      return IsHostResource<typename resources::get_resource<t_pol>::type,
+                            t_pol, SEGMENT, BODY>::invoke_forall(r, seg, body);
+    }
+
+    return dynamic_helper<IDX-1, POLICY_LIST>::invoke_forall(r, pol, seg, body);
+  }
+
+  template<typename SEGMENT, typename BODY>
+  static void invoke_forall(const int pol, SEGMENT const &seg, BODY const &body)
+  {
+    if(IDX==pol){
+      using t_pol = typename camp::at<POLICY_LIST,camp::num<IDX>>::type;
+      RAJA::forall<t_pol>(seg, body);
+      return;
+    }
+    dynamic_helper<IDX-1, POLICY_LIST>::invoke_forall(pol, seg, body);
+  }
+
+};
+
+template<typename POLICY_LIST>
+struct dynamic_helper<0, POLICY_LIST>
+{
+  template<typename SEGMENT, typename BODY>
+  static void invoke_forall(const int pol, SEGMENT const &seg, BODY const &body)
+  {
+    if(0==pol){
+      using t_pol = typename camp::at<POLICY_LIST,camp::num<0>>::type;
+      RAJA::forall<t_pol>(seg, body);
+      return;
+    }
+    RAJA_ABORT_OR_THROW("Policy enum not supported: ");
+  }
+
+
+  template<typename SEGMENT, typename BODY>
+  static resources::EventProxy<resources::Resource>
+  invoke_forall(RAJA::resources::Resource r, const int pol, SEGMENT const &seg, BODY const &body)
+  {
+    if(pol != 0) RAJA_ABORT_OR_THROW("Policy enum not supported: ");
+
+    using t_pol = typename camp::at<POLICY_LIST,camp::num<0>>::type;
+
+    return IsHostResource<typename resources::get_resource<t_pol>::type,
+                          t_pol, SEGMENT, BODY>::invoke_forall(r, seg, body);
+  }
+
+};
+
+
+  template<typename POLICY_LIST, typename SEGMENT, typename BODY>
+  void dynamic_forall(const int pol, SEGMENT const &seg, BODY const &body)
+  {
+
+    constexpr int N = camp::size<POLICY_LIST>::value-1;
```

Comment:
Would it be more clear if N was the whole size and we passed N-1 to dynamic_helper?

Reply:
Should we have a static_assert against size being 0?

Reply:
Yes on the static_assert, and sure I think either work.